examples/haiku_simple.py

```diff
@@ -0,0 +1,164 @@
+    """Loads the sub-sampled dataset as a generator of batches."""
+    ds = tfds.load("mnist:3.*.*", split=split).take(sample_size).cache().repeat()
+    if is_training:
+        ds = ds.shuffle(sample_size, seed=0)
+    ds = ds.batch(batch_size)
+    return iter(tfds.as_numpy(ds))
+
+
+def objective(trial):
+    # Make datasets.
+    train = load_dataset(
+        "train", is_training=True, batch_size=BATCH_SIZE, sample_size=N_TRAIN_SAMPLES
+    )
+    train_eval = load_dataset(
+        "train", is_training=False, batch_size=BATCH_SIZE, sample_size=N_TRAIN_SAMPLES
+    )
+    test_eval = load_dataset(
+        "test", is_training=False, batch_size=BATCH_SIZE, sample_size=N_VALID_SAMPLES
+    )
+
+    # Draw hyper-parameters
+    n_units_l1 = trial.suggest_int("n_units_l1", 4, 128)
+    n_units_l2 = trial.suggest_int("n_units_l2", 4, 128)
+    lr = trial.suggest_float("lr", 1e-5, 1e-1, log=True)
+
+    # Define feed-forward function by using sampled parameters
+    def net_fn(batch: Batch) -> jnp.ndarray:
+        """Standard MLP network."""
+        x = batch["image"].astype(jnp.float32) / 255.0
+        mlp = hk.Sequential(
+            [
+                hk.Flatten(),
+                hk.Linear(n_units_l1),
+                jax.nn.relu,
+                hk.Linear(n_units_l2),
+                jax.nn.relu,
+                hk.Linear(10),
+            ]
+        )
+        return mlp(x)
+
+    # Make the network and optimiser.
+    net = hk.without_apply_rng(hk.transform(net_fn))
+    opt = optax.adam(lr)
+
+    # Training loss (cross-entropy).
+    def loss(params: hk.Params, batch: Batch) -> jnp.ndarray:
+        """Compute the loss of the network, including L2."""
+        logits = net.apply(params, batch)
+        labels = jax.nn.one_hot(batch["label"], 10)
+
+        l2_loss = 0.5 * sum(jnp.sum(jnp.square(p)) for p in jax.tree_leaves(params))
+        softmax_xent = -jnp.sum(labels * jax.nn.log_softmax(logits))
+        softmax_xent /= labels.shape[0]
+
+        return softmax_xent + 1e-4 * l2_loss
+
+    # Evaluation metric (classification accuracy).
+    @jax.jit
+    def accuracy(params: hk.Params, batch: Batch) -> jnp.ndarray:
+        predictions = net.apply(params, batch)
+        return jnp.mean(jnp.argmax(predictions, axis=-1) == batch["label"])
+
+    @jax.jit
+    def update(
+        params: hk.Params,
+        opt_state: OptState,
+        batch: Batch,
+    ) -> Tuple[hk.Params, OptState]:
+        """Learning rule (stochastic gradient descent)."""
+        grads = jax.grad(loss)(params, batch)
+        updates, opt_state = opt.update(grads, opt_state)
+        new_params = optax.apply_updates(params, updates)
+        return new_params, opt_state
+
+    # Initialize network and optimiser; note we draw an input to get shapes.
+    params = net.init(jax.random.PRNGKey(42), next(train))
+    opt_state = opt.init(params)
+
+    best_test_accuracy = 0.0
+    # Train/eval loop.
+    for step in range(TRAIN_STEPS + 1):
```

Comment:
```suggestion
    for step in range(1, TRAIN_STEPS + 1):
```

I think without start=1, this script update parameters `TRAIN_STEPS + 1`, which sounds not so natural.